xapian-core/weight/pl2weight.cc

```diff
@@ -78,13 +78,52 @@ PL2Weight::init(double)
 
     // Calculate the lower bound on the weight.
     double P_min =
-	P1 + (wdfn_lower + 0.5) * log2(wdfn_lower) - P2 * wdfn_lower;
+	P1 + (wdfn_lower + 0.5) * log2(wdfn_lower) - P2 * wdfn_upper;
```

Comment:
I've committed a change to address this part fully - we now calculate upper_bound correctly whatever sign P2 has (dd6b359).

I've also dropped the lower_bound approach (which doesn't actually work as intended in general) and instead clamp the final weight to be >= 0 in get_sumpart() (1039fab).